nilearn/decomposition/base.py

```diff
@@ -0,0 +1,533 @@
+from sklearn.linear_model import LinearRegression
+from sklearn.utils import check_random_state
+from sklearn.utils.extmath import randomized_svd
+from sklearn.base import BaseEstimator
+from ..input_data import NiftiMapsMasker
+from ..input_data.masker_validation import check_embedded_nifti_masker
+from .._utils.cache_mixin import CacheMixin, cache
+from .._utils.niimg_conversions import check_niimg_4d
+from .._utils.niimg import _safe_get_data
+
+class MaskReducer(BaseEstimator):
+    """Mask and reduce provided data with provided masker, using a PCA
+
+    Uses a PCA (randomized for small reduction ratio) or a range finding matrix
+    on time series to reduce data size in time. For multiple image,
+    the concatenation of data is returned, either as an ndarray or a memorymap
+    (useful for big datasets that do not fit in memory).
+
+    Parameters
+    ----------
+    masker: NiftiMasker or MultiNiftiMasker
+        Masker to use to mask provided data
+
+    imgs: list of Niimg-like objects
+        See http://nilearn.github.io/building_blocks/manipulating_mr_images.html#niimg.
+        List of subject data
+
+    confounds: CSV file path or 2D matrix
+        This parameter is passed to signal.clean. Please see the
+        corresponding documentation for details.
+
+    reduction_ratio: 'auto' or float
+        How to reduce the data. If 'auto', reduce it to the provided number of
+        components. If float between 0. and 1., data will be reduced to the
+        provided ratio. Default: 'auto'
+
+    n_components: integer, optional
+        Number of components to be extracted by the PCA
+
+    random_state: int or RandomState
+        Pseudo number generator state used for random sampling.
+
+    memory_level: integer, optional
+        Integer indicating the level of memorization. The higher, the more
+        function calls are cached.
+
+    memory: joblib.Memory
+        Used to cache the function calls.
+
+    memory_mode: {'auto', 'array', 'memorymap'}
+        Whether to return a memorymap or and ndarray. Default: 'auto'
+
+    Retuns
+    ------
+    data: ndarray or memorymap
+        Concatenation of reduced data
+    """
+
+    def __init__(self, masker,
+                 reduction_ratio='auto',
+                 n_components=None, random_state=None,
+                 memory_level=0,
+                 memory=Memory(cachedir=None),
+                 mock=False,
+                 temp_folder=None,
+                 n_jobs=1):
+        self.masker = masker
+        self.reduction_ratio = reduction_ratio
+        self.n_components = n_components
+        self.random_state = random_state
+        self.memory_level = memory_level
+        self.memory = memory
+        self.temp_folder = temp_folder
+        self.mock = mock
+        self.n_jobs = n_jobs
+
+    def fit(self, imgs, confounds=None):
+
+        if not hasattr(imgs, '__iter__'):
+            imgs = [imgs]
+        else:
+            imgs = imgs
```

Comment:
@AlexandreAbraham : can we avoid this code thanks to our checkers? I don't like the idea of having it spread in the codebase.